src/node/rpc/member_frontend.h

```diff
@@ -330,6 +382,162 @@ namespace ccf
       return tx_modules->remove(name);
     }
 
+    void remove_jwt_keys(kv::Tx& tx, std::string issuer)
+    {
+      auto keys = tx.get_view(this->network.jwt_public_signing_keys);
+      auto key_issuer =
+        tx.get_view(this->network.jwt_public_signing_key_issuer);
+
+      key_issuer->foreach(
+        [&issuer, &keys, &key_issuer](const auto& k, const auto& v) {
+          if (v == issuer)
+          {
+            keys->remove(k);
+            key_issuer->remove(k);
+          }
+          return true;
+        });
+    }
+
+    bool set_jwt_public_signing_keys(
+      kv::Tx& tx,
+      ObjectId proposal_id,
+      std::string issuer,
+      const JsonWebKeySet& jwks)
+    {
+      auto issuers = tx.get_view(this->network.jwt_issuers);
+      auto keys = tx.get_view(this->network.jwt_public_signing_keys);
+      auto key_issuer =
+        tx.get_view(this->network.jwt_public_signing_key_issuer);
+
+      auto issuer_metadata_ = issuers->get(issuer);
+      if (!issuer_metadata_.has_value())
+      {
+        LOG_FAIL_FMT(
+          "Proposal {}: {} is not a valid issuer", proposal_id, issuer);
+        return false;
+      }
+      auto& issuer_metadata = issuer_metadata_.value();
+
+      remove_jwt_keys(tx, issuer);
+
+      // add keys
+      if (jwks.keys.empty())
+      {
+        LOG_FAIL_FMT("Proposal {}: JWKS has no keys", proposal_id);
+        return false;
+      }
+      bool did_add_key = false;
+      for (auto& jwk : jwks.keys)
+      {
+        if (keys->has(jwk.kid))
+        {
+          LOG_FAIL_FMT(
+            "Proposal {}: key id {} already added for different issuer",
+            proposal_id,
+            jwk.kid);
+          return false;
+        }
+        if (jwk.x5c.empty())
+        {
+          LOG_FAIL_FMT("Proposal {}: JWKS is invalid (empty x5c)", proposal_id);
+          return false;
+        }
+
+        auto& der_base64 = jwk.x5c[0];
+        auto der = tls::raw_from_b64(der_base64);
+
+        std::map<std::string, std::vector<uint8_t>> claims;
+        bool has_key_policy_sgx_claims =
+          issuer_metadata.key_policy.has_value() &&
+          issuer_metadata.key_policy.value().sgx_claims.has_value() &&
+          !issuer_metadata.key_policy.value().sgx_claims.value().empty();
+        if (
+          issuer_metadata.key_filter == JwtIssuerKeyFilter::SGX ||
+          has_key_policy_sgx_claims)
+        {
+          oe_verifier_initialize();
+          oe_verify_attestation_certificate_with_evidence(
```

Comment:
Check return code?

Reply:
If the cert isn't an OE attested cert or fails to parse for some other reason then the `claims` map will stay empty, which I use for all further checks. Should be enough?

Reply:
I believe it is safer to exit here/early if the attestation certificate verification fails?

Reply:
No, this works together with the key_filter. For MAA, the JWKS contains a mix of OE and non-OE certs and this code tries to extract the OE claims. If there are none, it's not an OE cert, which is fine.